### ISSUE TITLE
VSA and unrecognized fields

### ISSUE THREAD

**Issue by laurentsimon** [slsa-framework/slsa#899]
In the current VSA specs:
In parsing rules:
`Consumers MUST ignore unrecognized fields.`
In verifiedLevels:
`Users MAY add additional, non-SLSA properties to this field provided the values do not conflict with the definition of SlsaResult`

It would be useful to elaborate why we're being flexible on the use of unrecognized fields.

Something to keep in mind is what happens when a verifying tool encounters an unrecognized fields (that it does not understand). IIUC, verification will fail open. I don't know how much of a problem this is in practice for VSAs. 

In x509, they mark extensions as "critical": if a verifier encounters a critical extension that it does not understand, it must fail (fail safe). Is this relevant to VSAs?

It would also be useful to provide some guidance how verifiers / producers should name these (out of specs) fields, to avoid collision with (future) SLSA fields.

For example, you could imagine the VSA claims look as follow:
```json
"claims": [
  {
    "name": "slsa.dev/verifiedLevels"
    "critical": true,
    "value": _list_
  },
  {
    "name": "slsa.dev/resourceUri"
    "critical": false, // Verification is optional
    "value": "some-name"
  },
]
```

**Comment by MarkLodato** [comment]
The general guidance is that policies ought to be [monotonic](https://github.com/in-toto/attestation/blob/main/spec/v1/README.md#parsing-rules): ignoring an attestation or unrecognized field SHOULD NOT turn a "deny" into an "allow". In other words, policies SHOULD require existence of a positive rather than absence of a negative.

The docs do not currently say this for `verifiedLevels` but should. In other words, we should say that the consumer SHOULD check for existence of specific values and ignore anything it doesn't care about or recognize.

Do you have concerns with this approach?

Regarding `critical`, I'm concerned about added complexity. Could you provide a concrete use case for this?

**Comment by laurentsimon** [comment]
> 
> Do you have concerns with this approach?
> 
> Regarding `critical`, I'm concerned about added complexity. Could you provide a concrete use case for this?

yes, it adds some complexity. More of a general comment around `ignoring an attestation or unrecognized field SHOULD NOT turn a "deny" into an "allow" (`SHOULD` -> `MUST`?) in that it's basically describing critical fields that MUST NOT be ignored. If there are multiple tools that support SLSA verification, some tools won't understand certain custom fields: so even if the policy is designed to be monotonic, the tool *could* turn a DENY into an ALLOW. I'm taking the view that verification tools should be the developer's ally, not their enemy, so I'm trying to understand what the implications are for tool developers. A a tool user, I'd like my tool to warn me if I forget to pass important parameters for verification, instead of ignoring it. I think that's the high-level gist of my comment. But maybe those verification tools are policy specific and not SLSA tools per se?

**Comment by laurentsimon** [comment]
Here is an example I'd like to be able to avoid in the future https://github.com/argoproj/argo-cd/issues/14431

**Comment by kpk47** [comment]
> It would also be useful to provide some guidance how verifiers / producers should name these (out of specs) fields, to avoid collision with (future) SLSA fields.

We provide that guidance in the definition of SlsaResult: https://slsa.dev/verification_summary/v1#emslsaresult-stringem. Perhaps we should update verifiedLevels to be more explicit.

**Comment by laurentsimon** [comment]
> Not quite. x509's critical extensions fail verification if the verifier doesn't understand it.

yes. That's the concern I'm raising. Ignoring fields fails open. That's why x509 marks them as "critical" if they must be verified.

> x509-style critical fields may be harmful to SLSA, as they would complicate adding new values to slsaResult, which we intend to do every time we add a level or track.

I'm not proposing adding x509 fields. I'm saying that x509 deals with unknown fields a certain way that it fails safe, rather than fail open.

> I don't think the example from Argo applies here.

This example illustrates the problem of tooling not warning users. What we currently call "unknown values" will be discarded by tools that don't understand them. It fails open rather than fail safe.

**Comment by MarkLodato** [comment]
I don't understand. In the x509 model, the *producer* decides what's critical, thus a field in the message. Here, we're saying that the *consumer* decides what's critical, so a field doesn't make sense. A consumer ought to check for the presence of the things they care about. Any extra information MUST be irrelevant. It is against the spec for a producer to add an extension field (or another enum value) that invalidates the existing field. Does that make sense?

**Comment by kpk47** [comment]
Backing up a bit: it looks to me like VSAs ignore unrecognized fields because they're in-toto attestations, and ignoring unrecognized fields is required by the in-toto [parsing rules.](https://github.com/in-toto/attestation/blob/main/spec/v1/README.md#parsing-rules) @MarkLodato linked above. I hadn't made the connection until just now, so I'm pointing it out in case others also missed that point.

**Comment by laurentsimon** [comment]
Chatted with @kpk47 offline and maybe easier to discuss offline and summarize after.

@MarkLodato TL;DR: 
> A consumer ought to check for the presence of the things they care about

Agreed.

I think we're just looking at it from different angles: You're making the assumption that users will always do the right thing, I'm making the assumption that users will make mistakes and that tooling should help users prevent verification mistakes, e.g., https://github.com/argoproj/argo-cd/issues/14431. Being explicit about _what_ matters according to the producer is useful: the final consumer may ignore it _explicitly_, but ignoring seems opens the door to configuration mistakes.

The downside of being explicit is that if a consumer consumes a bunch of 3P dependencies, then every time one of the dependencies adds new "unknown" fields it will break the consumer. Another downside is addition of fields in VSA across non-major version updates.

Hope that clarifies a bit my concern.

**Comment by laurentsimon** [comment]
I don't think I've managed to convey exactly what I meant, but your responses have helped me clarify my though. I've managed to convince myself that the security guarantees I was after are out of scope of SLSA tooling. "Generic" SLSA tooling need not try to prevent advanced users (using custom or experimental fields) from making mistakes, as they'll use their own custom tooling anyway. 

Closing the issue. Thanks everyone!
